plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieCatalogModule.java

```diff
@@ -58,9 +62,53 @@ public static NessieIcebergClient createNessieIcebergClient(IcebergNessieCatalog
         icebergNessieCatalogConfig.getBearerToken()
                 .ifPresent(token -> builder.withAuthentication(BearerAuthenticationProvider.create(token)));
 
-        return new NessieIcebergClient(builder.build(NessieApiV1.class),
+        // default version is inferred by uri.
+        String apiVersion;
+        if (icebergNessieCatalogConfig.getClientAPIVersion().isPresent()) {
+            apiVersion = icebergNessieCatalogConfig.getClientAPIVersion().get();
+        }
+        else {
+            apiVersion = inferVersionFromURI(icebergNessieCatalogConfig.getServerUri().toString());
+        }
+
+        NessieApiV1 api;
+        switch (apiVersion) {
+            case "1":
+                api = builder.build(NessieApiV1.class);
+                break;
+            case "2":
+                api = builder.build(NessieApiV2.class);
+                break;
+            default:
+                throw new IllegalArgumentException(
+                        String.format(
+                                "Unsupported nessie client API version: %s. Can only be 1 or 2", apiVersion));
+        }
+
+        return new NessieIcebergClient(api,
                 icebergNessieCatalogConfig.getDefaultReferenceName(),
                 null,
                 ImmutableMap.of());
     }
+
+    private static String inferVersionFromURI(String uri)
```

Comment:
Ideally we should have exposed this from Iceberg repo
https://github.com/apache/iceberg/blob/2722290a7793dcb4f411aad271d5a22697c696f9/nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java#L136

But that needs another release (1.6.0, which may take a while)

Reply:
Return `ClientApiVersion.forName()`. This will throw if the API is unknown

Reply:
Didn't use `ClientApiVersion.valueOf()` but returning the `ClientApiVersion` from `inferVersionFromURI`